sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala

```diff
@@ -167,6 +167,24 @@ final class DataStreamReader private[sql](sparkSession: SparkSession) extends Lo
       className = source,
       options = extraOptions.toMap)
     ds match {
+      case s: MicroBatchReadSupport =>
+        val tempReader = s.createMicroBatchReader(
+          java.util.Optional.ofNullable(userSpecifiedSchema.orNull),
+          Utils.createTempDir(namePrefix = s"temporaryReader").getCanonicalPath,
+          options)
+        // Generate the V1 node to catch errors thrown within generation.
+        try {
+          StreamingRelation(v1DataSource)
+        } catch {
+          case e: UnsupportedOperationException
```

Comment:
https://github.com/apache/spark/blob/9a2b65a3c0c36316aae0a53aa0f61c5044c2ceff/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L266

I agree that it would be nice to change this exception, but I don't know whether we can.

Reply:
On reflection, there's actually a better way to do this which does not need to use exceptions as control flow. I didn't notice before because lookupDataSource returns Class[_] for some reason.